frontend/catalyst/compilation_pipelines.py

```diff
@@ -550,9 +546,87 @@ def __call__(self, *args, **kwargs):
         else:
             args = CompiledFunction.promote_arguments(self.c_sig, r_sig, *args)
 
+        if any(isinstance(arg, jax.core.Tracer) for arg in args):
+            # Only compile a derivative version of the compiled function when needed.
+            if self.jaxed_qfunc is None:
+                self.jaxed_qfunc = JAX_QJIT(self)
+
+            return self.jaxed_qfunc(*args, **kwargs)
+
         return self.compiled_function(*args, **kwargs)
 
 
+class JAX_QJIT:
+    """Wrapper class around :class:`~.QJIT` that enables compatibility with JAX transformations.
+
+    The primary mechanism through which this is effected is by wrapping the invocation of the QJIT
+    object inside a JAX ``pure_callback``. Additionally, a custom JVP is defined in order to support
+    JAX-based differentiation, which is itself a ``pure_callback`` around a second QJIT object which
+    invokes :func:`~.grad` on the original function. Using this class thus incurs additional
+    compilation time.
+
+    Args:
+        qfunc (QJIT): the compiled quantum function object to wrap
+    """
+
+    def __init__(self, qfunc):
+        @functools.wraps(qfunc, assigned=("__annotations__",), updated=())
+        def deriv_wrapper(*args, **kwargs):
+            # Only consider differentiable arguments (i.e. of type floating point).
+            all_diff_args = [i for i in range(len(args)) if args[i].dtype.kind == "f"]
+
+            return catalyst.grad(qfunc, argnum=all_diff_args)(*args, **kwargs)
+
+        deriv_wrapper.__name__ = "deriv." + qfunc.__name__
```

Comment:
Could this be expected to be the entry-point to the compiled function?

Reply:
If I understand you correctly, yes. The `deriv_wrapper` is directly passed to `QJIT` in the next step to instantiate a new compiled function, making it the entry point of that newly compiled function.

Reply:
I think we need to phase the `.` in `deriv.${name}` :sweat_smile: otherwise, linking the wrapper and catching the exception will involve a linker script.

Reply:
To be more concrete, substitute the dot for an underscore.

Reply:
Sounds good to me :)

Reply:
https://github.com/PennyLaneAI/catalyst/pull/96/commits/952d2eab74aa15b82be4f58f10b0f33ab0fc2b4e